### ISSUE TITLE
[HW/CI] remove CSR tests from CI

### ISSUE THREAD

**Issue by rasmus-madsen** [lowRISC/opentitan#7494]
Currently the CSRs are part of the CI.
while I do find it a great Idea to have the tests added by the IP owner in the CI, I don't think we need the CSRs

not only does the CSR's slow down CI
I also don't see any value add - there is no risk of compromising the master branch with a failing CSR.
but it will hold up DV work - having to go fix CSRs test cases that may be irrelevant in the early phases of V1.

I realize there are workarounds for the DV work, but I would still like the reduced CI run time.
if we decide that CSRs should stay in CI - we should update the checkpoints to make passing CSR a V1 requirement!

Maybe we can make this an item on the next DV meeting

**Comment by sriyerg** [comment]
TL; DR; my vote is to keep the CSR tests, audit the public CI instead and continue profiling the DV env for bottlenecks. 

Thanks @rasmus-madsen for raising this issue. FYI, having all CSR tests pass is already a V1 requirement. 
https://docs.opentitan.org/doc/project/checklist/#v1

For any new block, the CSR tests come for free as soon as the initial testbench is autogenerated - and as such serve as a pretty good starting point. The advantage of adding them to CI is if the designers update the HJSon later, the CSR tests help point out mistakes / assumptions made in the HJSon very early on. IMO that is valuable and saves quite a lot of future DV debug overhead.

Depending on how soon a DV resource is allocated, there may be a significant amount of time until the first functional smoke test is developed. The CSR tests serve as a pretty good presubmit check in the meantime. 

CI failures due to CSRs requiring exclusions implicitly help confirm that the CSRs are exhibiting the desired behavior, much before functional tests are developed. Especially, when CSRs are added / deleted during D1 -> D3. 

Our (loose) CI runtime requirement is <25 minutes. We are averaging between 16-19 minutes, which IMO is pretty good considering it runs 89 tests spanning all of our DUTs. The much much bigger bottleneck is the public CI (1-2hr runtime) which runs way too many things. We ought to be auditing that instead. Reducing the private CI runtime will likely not yield any benefit.

There are other ways of reducing CI runtime - we have done a few profiling exercises this year. We have improved the runtime by eliminating bottlenecks (refactored code, removed constructs there were not useful) and have seen speed up of as much as 40%. IMO we should all continue doing that once a month.

**Comment by weicaiyang** [comment]
I also prefer to keep the IP-level CSR tests, but perhaps remove the `chip_csr_rw`.

As Sri mentioned, running CSR tests helps us to capture some mistakes due to some updates the Hjson or design. Designer can make a quick fix or adjust CSR exclusion. Now designers know how to debug all kinds of CSR tests and use correct CSR exclusion. Without running those in CI, DV may end up to spending more time on debugging/fixing CSR test failures. 

The reason I prefer to remove `chip_csr_rw` is that the passing rate of this test is often very low (only 20% from today's nightly result). Some block may not include csr test into smoke, but `chip_csr_rw` will automatically include it when it's added into top_earlgrey. And when it fails, it's also hard to know if the failure is due to my update. 
For example, if I update block1, but `block2_csr_rw` fails. This is very obvious that it's not due to my update and I can ping the owner of block2 to take a look. Or I can temporarily remove `block2_csr_rw` from smoke to unblock the PR if we need more time fix that. But if it's chip_csr_rw fails, I may need to take a close look to know what the root cause could be. After a long debug, we may find the root cause is due to some issues in block2, block3 and block4. It's just very unfortunate that block2/3/4_csr_rw pass with seed=1, but chip_csr_rw fail.

Thanks @rasmus-madsen for raising this. If we don't have a conclusion by this week, I will add it into our next DV meeting.

**Comment by sriyerg** [comment]
I agree with the idea of removing `chip_csr_rw` test from presubmit CI, but unfortunately there is no easy way to selectively remove it for chip but keep it for the blocks.

**Comment by rasmus-madsen** [comment]
@sriyerg @weicaiyang I think I I got a leg in both camps here.
not sure what types of errors in the Hjson would be caught by the CSR's and not the smoke test?
also, CSR tests are mostly relevant for design engineers - the DV engineer does not have any contact with them other than making sure they are not accidentally running tasks from the IP DV.
and there is nothing more frustrating than a CI that fails due to something unrelated.
the suggested workaround is to exclude the register that fails.
That just seem like a dangerous path

@weicaiyang I do not agree with your statement 

> DV may end up to spending more time on debugging/fixing CSR test failures.

if they are still in regression they should show up within 24 hours.
at the moment I almost have the feeling we are using CI as a regression.

from my point of view the listed pros are not really that big, and certainly doesn't outweigh the cons
with slow CI speed - and failing CI

@rswarbrick @senelson7 @matutem 
any thoughts?

**Comment by rswarbrick** [comment]
I'm not sure what I think about this. Honestly, I'm happy with a super-finicky CI, even if it means a couple of PRs get slightly delayed.

About this comment:
> at the moment I almost have the feeling we are using CI as a regression.

Yeah we are, sort of. Maybe that's unavoidable in a cross-organisation project (because no-one at Google can shout at me if I break stuff, it's better to make sure I don't break it in the first place :smile:)

**Comment by sriyerg** [comment]
@rasmus-madsen, please note that the CSR tests are developed in DV / UVM environment, they are added to DV testplan and hence, it is absolutely the DV engineer's responsibility to maintain the health. By adding them to CI, we end up saving the DV engineer's time by splitting that maintenance overhead with the designers. Isn't that useful (that is what Weicai was alluding to in his earlier comment)?

The exclusion is applied to a register if the access check fails ONLY if it is determined on examination of the behavior through waves, that it is safe to do so (CSR tests are only meant to ensure SW can fully access the IP's register space - they are not meant to predict external HW writes). If there is evidence of exclusions being applied blindly for the sake of getting the test to pass, then please report! That is definitely not the right way to go. 

You feel we are using CI to run regression. Take AES for example - CI (private) runs a mere 3 tests:
https://dev.azure.com/lowrisc/opentitan-private/_build/results?buildId=47784&view=logs&j=a108fe5f-1aa1-5d88-0e90-2681f9ae22bc&t=fb9e6058-94ab-5f9a-0d29-b4f23fc20bc4&l=403

The full nightly runs 742 tests for AES:

I am not sure I understand your insistence of removing the CSR test

**Comment by rasmus-madsen** [comment]
@sriyerg I am trying to review what we are doing and make sure it makes sense.
the FGPA builds are another thing that I feel could be removed from CI  or limited to compile and elaboration.

as mentioned we are really using CI for regression - which IMO is at best, an unneeded overhead.

Anyway, it seems between the people who felt like chiming in - the consensus is that the current way works well.
which is think is probably a good steering rod.
So let's keep the CSRs in.
unless there are any more comments I am ok with closing this issue and leaving it as is!

**Comment by sriyerg** [comment]
One more problem (likely the most significant) is the RAL tooling is still getting significantly updated to this day (and will likely continue to receive more updates in future), where not just the tooling but also the RTL is getting modified. CSR tests in CI serve as a good check to ensure we are not leaving things in a broken state. The smoke tests IMO are not extensive enough to cover an IP's complete SW interface.

**Comment by sriyerg** [comment]
> unless there are any more comments I am ok with closing this issue and leaving it as is!

Sure, that sounds good. Though, your comment above is a bit concerning: 
> the suggested workaround is to exclude the register that fails.
> That just seem like a dangerous path

Please do let us know if there were cases where exclusions were added without properly analyzing the failure.

**Comment by rasmus-madsen** [comment]
> Please do let us know if there were cases where exclusions were added without properly analyzing the failure.

I don't know of any - at the time where I started thinking about this, someone suggested excluding the failing register to get CI to pass.
Which IMO is not a valid path to take - since this exclusion could be forgotten.
